p2p/pex/reactor.go

```diff
@@ -76,31 +129,149 @@ func (r *ReactorV2) OnStop() {
+// Envelope messages from the pexCh.
+func (r *ReactorV2) processPexCh() {
+	defer r.pexCh.Close()
+	defer r.peerUpdates.Close()
+
+	for {
+		select {
+		case <-r.closeCh:
+			r.Logger.Debug("stopped listening on PEX channel; closing...")
+			return
+
+		// handle peer updates
+		case peerUpdate := <-r.peerUpdates.Updates():
+			r.processPeerUpdate(peerUpdate)
+
+		// outbound requests for new peers
+		case <-r.waitUntilNextRequest():
+			r.sendRequestForPeers()
+
+		// inbound requests for new peers or responses to requests sent by this
+		// reactor
+		case envelope := <-r.pexCh.In:
+			if err := r.handleMessage(r.pexCh.ID, envelope); err != nil {
+				r.Logger.Error("failed to process message", "ch_id", r.pexCh.ID, "envelope", envelope, "err", err)
+				r.pexCh.Error <- p2p.PeerError{
+					NodeID: envelope.From,
+					Err:    err,
+				}
+			}
+		}
+	}
+}
+
 // handlePexMessage handles envelopes sent from peers on the PexChannel.
 func (r *ReactorV2) handlePexMessage(envelope p2p.Envelope) error {
 	logger := r.Logger.With("peer", envelope.From)
 
-	// FIXME: We may want to add DoS protection here, by rate limiting peers and
-	// only processing addresses we actually requested.
 	switch msg := envelope.Message.(type) {
+
+	// LEGACY PEX MESSAGES
 	case *protop2p.PexRequest:
-		pexAddresses := r.resolve(r.peerManager.Advertise(envelope.From, maxAddresses), maxAddresses)
+		// check if the peer hasn't sent a prior request too close to this one
+		// in time
+		if err := r.markPeerRequest(envelope.From); err != nil {
+			return err
+		}
+
+		// parse and send the legacy PEX addresses
+		pexAddresses := r.resolve(r.peerManager.Advertise(envelope.From, maxAddresses))
 		r.pexCh.Out <- p2p.Envelope{
 			To:      envelope.From,
 			Message: &protop2p.PexResponse{Addresses: pexAddresses},
 		}
 
 	case *protop2p.PexResponse:
+		// check if the response matches a request that was made to that peer
+		if err := r.markPeerResponse(envelope.From); err != nil {
+			return err
+		}
+
+		// check the size of the response
+		if len(msg.Addresses) > int(maxAddresses) {
+			return fmt.Errorf("peer sent too many addresses (max: %d, got: %d)",
+				maxAddresses,
+				len(msg.Addresses),
+			)
+		}
+
 		for _, pexAddress := range msg.Addresses {
+			// no protocol is prefixed so we assume the default (mconn)
 			peerAddress, err := p2p.ParseNodeAddress(
 				fmt.Sprintf("%s@%s:%d", pexAddress.ID, pexAddress.IP, pexAddress.Port))
 			if err != nil {
-				logger.Debug("invalid PEX address", "address", pexAddress, "err", err)
 				continue
 			}
-			if err = r.peerManager.Add(peerAddress); err != nil {
-				logger.Debug("failed to register PEX address", "address", peerAddress, "err", err)
+			added, err := r.peerManager.Add(peerAddress)
+			if err != nil {
+				logger.Error("failed to add PEX address", "address", peerAddress, "err", err)
+			}
+			if added {
+				r.newPeers++
+				logger.Debug("added PEX address", "address", peerAddress)
+			}
+			r.totalPeers++
```

Comment:
We'll need mutexes if we have separate threads for peer update and message processing work.